### ISSUE TITLE
Add Collate extension to `Sql`

### ISSUE THREAD

**Issue by Corey-M** [linq2db/linq2db#2729]
Recently ran into a problem where an app I wrote to do some database merges failed at a client's site due to a collation mismatch. When attempting to merge records from a temporary table the merge failed with a collation mismatch error.

After much searching it appears that there is currently no way to control collation of string expressions in Linq2DB. I've written a simple extension that does what I need:

```C#
using LinqToDB.SqlQuery;

public static class SqlExtensions
{
    [Sql.Extension("{expr} COLLATE {collation}", ServerSideOnly = true, BuilderType = typeof(NamedCollationBuilder))]
    public static string Collate(this string expr, string collation)
        => throw new InvalidOperationException("Collation is server-side only.");
    
    internal class NamedCollationBuilder : Sql.IExtensionCallBuilder
    {
        public void Build(Sql.ISqExtensionBuilder builder)
        {
            var expr = builder.GetExpression("expr");
            var collation = builder.GetValue<string>("collation");
            
            builder.ResultExpression = new SqlExpression($"{{0}} COLLATE {collation}", Precedence.Primary, expr);
        }
    }
}
```

Usage:

```C#
var query = 
    from l in Left
    join r in Right on l.SomeField equals r.AnotherFIeld.Collate("SQL_Latin1_General_CP1_CS_AS")
    select 1;
```

Unlike the `DateParts` enumeration used in `Sql.DatePart` there is no list of values that can be made coherent between database engines so I don't see a simple way to restrict the `collation` parameter by enum, etc. and so far I've not found a handy mechanism for forcing this to accept only client-side variables and constants for the `collation` parameter.

### Environment details
linq2db version: All
Database Server: Most
Operating system: All
.NET Framework: All

**Comment by viceroypenguin** [comment]
This looks good to me. I've thought about this a time or two. Since the code is written, would you mind creating a PR w/ this method and some unit tests?

**Comment by sdanyliv** [comment]
I think `Sql.Expression` is enough with `InlineParameters = true`

**Comment by Corey-M** [comment]
@sdanyliv As far as I can tell `InlineParameters = true` renders the `collation` parameter as an inline SQL string, which does not work with the `COLLATE` clause. It has to pass the collation name as literal text into the SQL query string. We need `COLLATE SQL_Latin1_General_CP1_CS_AS` rather than `COLLATE N'SQL_Latin1_General_CP1_CS_AS'`. The only way I've found to do that is to use a builder class.

I've considered a few ways to do this and so far the code above is the simplest. I can add a simple regex to check for invalid characters (no spaces, etc). Will put something together for a PR and get some feedback.

And no, I have no idea how to write a test case for this.

**Comment by jogibear9988** [comment]
Maybe we should add a Function for each Database Driver with a enum for each of them as an additional extension?

**Comment by Corey-M** [comment]
@jogibear9988 I considered it. Even built one for SQL Server 2017's 3,955 different collations... then discovered that SQLite has APIs to create new collations, PostgreSQL and Firebird (at least) have `CREATE COLLATION` statements, collation lists change between server versions... it became a little silly. Oh, and Access doesn't support a `COLLATE` clause, so that's a thing.

I mean, yeah, I could add an `SQLServer2017Collation` enumeration, method and builder class. It's not even that much effort since I can query the list straight off my server. I could probably find other SQL server instances to get the collation lists from, add the 3 basic collations for SQLite etc. That would only cover the statics for two of the supported database engines, we'd still need the collation-as-string form to support dynamic collations and other database engines until someone adds them to the pile.

Ultimately this seems like a handy support feature for those oddball cases like mine where the DB schema uses inconsistent collation or the server's default schema is different to the schema used in the database. Not sure it's worth the effort to try to support all of the possibilities through enums and such... but that's not my call I guess.

**Comment by Corey-M** [comment]
So... anyone got any ideas on how to add a unit test for this?

**Comment by pebezo** [comment]
@Corey-M Thanks for that example! It helped me solve and issue with PostgreSQL (without having to use stored procs). One peculiar thing with PG is that the collation has to be wrapped in double quotes, not single quotes. Using it like `Field.Collate("\"en_US\"")` works though. It also works to chain it (which is super cool) like `where Field.Collate("\"en_US\"").Contains("something")`

**Comment by sdanyliv** [comment]
@pebezo, for PostgreSQL extension should be extended and it is not needed to manually escape collation name.
```cs
public static class SqlExtensions
{
    [Sql.Extension("{expr} COLLATE {collation}", ServerSideOnly = true, BuilderType = typeof(NamedCollationBuilder))]
    [Sql.Expression(ProviderName.PostgreSQL, "{expr} COLLATE {collation}", ServerSideOnly = true, InlineParameters = true))]
    public static string Collate(this string expr, string collation)
        => throw new InvalidOperationException("Collation is server-side only.");
    
    internal class NamedCollationBuilder : Sql.IExtensionCallBuilder
    {
        public void Build(Sql.ISqExtensionBuilder builder)
        {
            var expr = builder.GetExpression("expr");
            var collation = builder.GetValue<string>("collation");
            
            builder.ResultExpression = new SqlExpression($"{{0}} COLLATE {collation}", Precedence.Primary, expr);
        }
    }
}
```

**Comment by MaceWindu** [comment]
>> anyone got any ideas on how to add a unit test for this?

1. check generated sql
2. or better: test should perform collation-dependent operation like string comparison with couple of collations with different outcome

**Comment by pebezo** [comment]
@sdanyliv At least with PG the collation does not have to be quoted all the time. For example, if you create a custom collation (myschema.ci) then this query is valid:
``` 
select * from myschema.mytable p where p.name collate myschema.ci = 'AbCd'
```
but if you want to use one of the built-in one, you'd have to write:
``` 
select * from myschema.mytable p where p.name collate "C.UTF-8" = 'AbCd'
```
While ugly, it may be better to just for the double quotes at usage...

**Comment by tolfino** [comment]
I tried using the example to do dynamic collation, but it looks like Linq2DB will cache the first generation of the expression and then all further uses will use the cached expression: i.e. if `en-x-icu` is passed as the collation to `Collate()`, a second call with e.g. `collate: zh-Hans-x-icu` will continue using `en-x-icu`.

**Comment by tolfino** [comment]
Annotating `collation` with `[SqlQueryDependent]` fixes it :)

**Comment by MaceWindu** [comment]
Fixed by #2731
